### PR TITLE
feat: add vips straighten

### PR DESCRIPTION
**Pull request by knarewski**
## Background
GdkPixbuf-based image processing can take over 1GB of memory for processing a single high-resolution photo. We've decided to introduce another processor, hoping that it'll improve Morandi's memory profile and performance. 

## Problems
Vips processor doesn't support straighten operation

## Solution
Add support for straighten operation

## Notes
- Visually all the images appear to be slightly shifted. Since I didn't spot any issues with maths, I'm inclined to assume that it's due to rounding/trimming during computations
- Because of the arbitrary angle rotation's nature, the post-rotation image is bigger; the "empty" areas in the corners are filled with colour; because of that, the rotated image is cropped and them zoomed to preserve original dimensions; the background colour, however, has a tendency to slightly bleed on the edges; that tendency is more visible for vips processor; in order to reduce that effect, I've changed the background colour in vips to gray; see screenshots below for example.

## Pixbuf vs Vips comparison
| image |
|:--:|
| Chunk of the test image rotated by arbitrary angle, before cropping and zooming (vips); note the gray colour which blends with the image edges |

|   |
|:--:|
| Close-up (4000%) of background bleed on the white image (left - pixbuf, right - vips) |

|   |
|:--:|
| straighten operation (-20 degrees) (left - pixbuf, right - vips), mean average error: `0.01176` |

Other images can be compared by opening files with identical names in `spec/fixtures/reference_images/` and `spec/fixtures/reference_images/vips/`

### CODE REVIEW

File: spec/morandi_spec.rb [hunk 1]
```diff
@@ -239,7 +239,7 @@
       end
     end
 
-    describe 'when given a straighten option', vips_wip: processor_name == 'vips' do
+    describe 'when given a straighten option' do
```
Comment (MrLukeSmith) [inline spec/morandi_spec.rb:242]: "Consider testing angles that are multiples of 45" - GPT

**Review by MrLukeSmith — COMMENTED** [review]
This exceeds my mathematical capabilities. I'll defer to @Surge9 for the yay / nay approval. He's more fond of maths, IIRC. 😅

But I ran the class through GPT and it raised a few points that might be of value.

File: lib/morandi/operation/vips_straighten.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        post_scale_bounding_box_height = post_rotation_bounding_box_height * scale
+
+        width_diff = post_scale_bounding_box_width - original_width
+        height_diff = post_scale_bounding_box_height - original_height
+
+        # Round to nearest integer to reduce risk of ROTATION_BACKGROUND_FILL_COLOUR being visible in the corner
+        crop_x = (width_diff / 2).round
+        crop_y = (height_diff / 2).round
+
+        img.crop(crop_x, crop_y, original_width, original_height)
```
[the remark this answers is not in the document]
Comment (knarewski) [reply lib/morandi/operation/vips_straighten.rb:56]: Rounding using "normal" mathematical rules yielded the most reliable results to me (the gray bleed was most subtle and even on both sides)

File: lib/morandi/operation/vips_straighten.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        original_width = img.width
+        original_height = img.height
+
+        # It is possible to first rotate, then fetch width/height of resulting image to calculate scale,
+        # but that would make us lose precision which degrades cropping accuracy
+        rotation_value_rad = angle * (Math::PI / 180)
+        post_rotation_bounding_box_width = (img.height.to_f * Math.sin(rotation_value_rad).abs) +
+                                           (img.width.to_f * Math.cos(rotation_value_rad).abs)
+        post_rotation_bounding_box_height = (img.width.to_f * Math.sin(rotation_value_rad).abs) +
```
[the remark this answers is not in the document]
Comment (knarewski) [reply lib/morandi/operation/vips_straighten.rb:34]: See https://github.com/livelink/morandi-rb/pull/39#discussion_r1871489756

**Review by MrLukeSmith — APPROVED** [review]
Thanks for entertaining the AI critique. I appreciate it's not your favourite approach, but it made it possible for me to better understand the specifics and the few points raised felt plausibly relevant. 😉

It was intended to be a knowledge filler for me as opposed to a wild goose chase for you. Apologies if it came across differently. 🙇

The comprehensive testing coupled with the context provided LGTM. 

Feel free to wait for a second opinion if desired.

**Pull request merged**